API Nonces: Runtime cleanup

sergeymitr opened the pull request:
#### Changes proposed in this Pull Request:
- The Jetpack Nonce functionality is extracted from the Connection Manager, and moved into a separate class.
- The new class `Nonce_Handler` is covered with unit tests.
- The Jetpack Nonces are getting removed right after new nonce gets added, no longer requiring WP-cron to be working.

#### Jetpack product discussion
- `p1598036456118800-slack-CBG1CP4EN`
- `p9dueE-1Oi-p2`

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
1. Connect Jetpack.
2. Run `composer install` to refresh the classmap.
3. Disable WP-cron to concentrate on testing the runtime cleanup: `define( 'DISABLE_WP_CRON', true );`
4. Use the Jetpack Debug's "Mocker" tool to create a bunch of old nonces (a few dozens would be enough, unless you want to test the performance too).
5. Count the nonces, write down how many are there:
```sql
SELECT COUNT(*) FROM wp_options WHERE option_name LIKE 'jetpack\_nonce\_%';
```
6. Go to the [Developer Console](https://developer.wordpress.com/docs/api/console/) and send an API request:
`WP.COM API`, `GET`, `v1.2`, `/sites/your-site.com`
7. Run the `COUNT` SQL query again and confirm that the number of nonces decreased by 9 (1 new nonce, 10 old removed).
8. Run the API requests until all the old nonces get removed. The new nonces will not get removed, because they are created less than an hour ago.
9. When the nonces will no longer get removed (and start to grow in number instead), fetch the oldest:
```sql
SELECT FROM_UNIXTIME(option_value, '%Y-%m-%d %H:%i:%s') FROM wp_options WHERE option_name LIKE 'jetpack\_nonce\_%' ORDER BY option_value LIMIT 1;
```
10. Convert the UTC date to your local timezone, and make sure that it's not older than an hour.
11. Disconnect Jetpack. Confirm that all existing nonces got removed, including the fresh ones:
```sql
SELECT COUNT(*) FROM wp_options WHERE option_name LIKE 'jetpack\_nonce\_%';
```

#### Proposed changelog entry for your changes:
- Switching Jetpack nonces cleanup from WP Cron to runtime.

sergeymitr committed:
API Nonces: Refactoring and unit tests.

- The Jetpack Nonce functionality is extracted from the Connection Manager, and moved into a separate class.
- The new class `Nonce_Handler` is covered with unit tests.

sergeymitr committed:
API Nonces: Runtime cleanup.

- Every time a nonce is added, we remove two nonces older than an hour.
- Two new filters: one to prevent the cleanup, another to change how many nonces to delete.
- Unit tests.

sergeymitr committed:
API Nonces: Database performance optimization.

- The `DELETE` SQL query is split in two: a `SELECT` to load `option_id` values, and a `DELETE` by the primary keys.
- Unit tests adjusted to accomodate the change.

sergeymitr committed:
API Nonces: Refactoring of the "full cleanup" functionality.

The old functionality to remove all existing nonces can't function when huge number of nonces are present. The SQL query locks the options table, and takes very long time to complete.

This commit rewrites the functionality:
- The "full cleanup" SQL query was split in two.
- The `Nonce_Handler` class is refactored to avoid code duplication.
- Method `Jetpack::clean_nonces()` is removed (it's been deprecated since 7.7.0).

sergeymitr committed:
API Nonces: Checking if the table is locked.

If there is a non-optimized SQL query running over the options table, it may lock the existing records, making the runtime nonce cleanup wait for a long time before timing out.
This commit introduces a check if the table is in use before running the runtime cleanup.

anomiex commented on packages/connection/src/class-nonce-handler.php:
```diff
@@ -0,0 +1,197 @@
+			$return = false;
+		}
+
+		$wpdb->show_errors( $show_errors );
```
Do we care about the possibility of anything between line 59 and here throwing, and so preventing the show_errors from being reset?

sergeymitr replied:
Good catch, thanks for forcing me to dig this up :)

This piece of history was added 9 years ago and released in `Jetpack 1.2.3`, fascinating stuff: https://github.com/Automattic/jetpack/commit/51fe166af6a968db3c65b38aa56b8e41c903d439

Since then it traveled all over Jetpack:
- `jetpack.php`
- `class.jetpack.php`
- `class-manager.php`
- and now we're moving it into `class-nonce-handler.php`

I'm guessing it was added to gracefully catch the "duplicate entry" error before we added the `SELECT` there, so it was kind of two requests in one: `INSERT` checked if the entry exists, and added it if it didn't.

On the other hand, it seems like hiding errors is still relevant today in case of a race condition. If two requests with the same nonce come simultaneously, that nonce may be added between the `SELECT` and `INSERT`, making `INSERT` throw the "duplicate entry" error.

So I would say we'd better leave it there.

anomiex replied:
I think you answered a different question than what I was asking. But that was interesting, thanks for the explanation!

I was just wondering whether we should have a `try { ... } finally { $wpdb->show_errors( $show_errors ); }` or something like that, or if we expect that any exception thrown here is going to be uncaught or caught by a fatal error handler of some sort so the show_errors flag remaining false wouldn't be a problem.

anomiex replied:
As for the possible race leading to a "duplicate entry" error, there's a simple enough solution for that: use `INSERT IGNORE`, which will have MySQL specifically avoid raising the error for the INSERT. Some quick testing suggests that `$wpdb->query()` will return 0 for an ignored insert, versus 1 for a successful insert and false for a duplicate entry without IGNORE.

You could also use a transaction, and `SELECT ... FOR UPDATE` in the original select. But MySQL can get a bit funny about the gap locks involved when using `SELECT ... FOR UPDATE` to guard an INSERT: if two processes both lock the same gap with `SELECT ... FOR UPDATE`, then their INSERTs will block each other even if the inserts wouldn't actually conflict. Sigh.

sergeymitr replied:
Good idea, I added that 👍

anomiex commented on packages/connection/src/class-nonce-handler.php:
```diff
@@ -0,0 +1,197 @@
+	 * @return true
+	 */
+	public static function clean_all( $cutoff_timestamp = PHP_INT_MAX ) {
+		for ( $i = 0; $i < 1000; ++$i ) {
```
Why 1000?

I guess the idea is to not run for a very long time if there are millions of nonces, but it does seem to belie the stated contract that this deletes *all* nonces.

sergeymitr replied:
I think it's unrealistic to expect over a million of nonces to be removed at one request.
That way the person will never be able to disconnect Jetpack.

Good catch though, I'll add a comment mentioning that limit.

zinigor replied:
I just ran into a disconnect request timeout because of this function. The `clean_all` run would have probably caused the same problem on an old code. Since it's the last chance to remove everything, could we rely on the index and instead do something like:
```
DELETE FROM wp_options WHERE `option_name` >= 'jetpack_nonce_' AND `option_name` < 'jetpack_nonce_ZZZZZZZZZZZZZZZZZZZZZZZZZZZZ'
```

zinigor replied:
Just ran some tests, and it doesn't look good for sites with more than about 500k records:
```
MariaDB [wordpress_70]> DELETE FROM wp_options WHERE `option_name` >= 'jetpack_nonce_' AND `option_name` < 'jetpack_nonce_ZZZZZZZZZZZZZZZZZZZZZZZZZZZZ';
Query OK, 5000 rows affected (0.234 sec)

MariaDB [wordpress_70]> DELETE FROM wp_options WHERE `option_name` >= 'jetpack_nonce_' AND `option_name` < 'jetpack_nonce_ZZZZZZZZZZZZZZZZZZZZZZZZZZZZ';
Query OK, 1790000 rows affected (11 min 22.798 sec)
```

zinigor replied:
Right, what I'm mostly concerned about is the timeout when disconnecting. We try to clean up everything as much as possible, but in case of a site with many nonces stored, this may cause a timeout. We don't really have a good option here as far as I can see. We're going to leave the site with leftover notices anyway, but we can at least try to clean up as much as we can. The million entries that the method is limited to right now in case of a disconnect is not going to work, I think we need some sort of time constraint here to avoid timeouts. What do you think @sergeymitr ?

sergeymitr replied:
@zinigor
I like the idea.
Since default maximum execution time is 30 seconds, we could limit it to 20 seconds in addition to existing limit (1k iterations 1k records each).

So if not all records are removed in 20 seconds, we can't really do much with it. We could add wp-cron cleanup job specifically for this use case, but I don't think it's worth the added complexity. If somebody disconnects Jetpack, they're most likely going to disable or remove it anyway.

Something like this:
```suggestion
		for ( $time = time(), $i = 0; $i < 1000 && time() > $time - 20; ++$i ) {
			// ...
		}
```

sergeymitr replied:
I added the time limit: https://github.com/Automattic/jetpack/pull/17030/files#diff-c856b8c22febf99e6f16144e03d1d277R109

anomiex commented on packages/connection/src/class-nonce-handler.php:
```diff
@@ -0,0 +1,197 @@
+
+	/**
+	 * Check if the options table is locked.
+	 *
```
Worth acknowledging in the docs that use of this is subject to races, but we don't care since it's enough to mostly work around a long-query issue from some other plugin?

sergeymitr replied:
Yes, good point, I'll add that to the comments.

anomiex commented on packages/connection/src/class-nonce-handler.php:
```diff
@@ -0,0 +1,197 @@
+		$ids = $wpdb->get_col(
+			$wpdb->prepare(
+				"SELECT option_id FROM `{$wpdb->options}`"
+				. " WHERE `option_name` >= 'jetpack_nonce_' AND `option_name` < %s"
```
There's a logic error here, but it's unlikely to affect anything in practice for about 266 years.

If someone tries to pass a `$cutoff_timestamp` before 2001-09-09T01:46:40Z, it'll also delete most timestamps after that date. Say they want to delete all nonces before the year 2000, so they pass 946684800. But since as a string comparison "jetpack_nonce_1000000000" < "jetpack_nonce_946684800"...

The same sort of thing could happen in 2286 when timestamps reach 11 decimal digits.

But since WordPress didn't exist until 2003 (so there's no sane reason to set an earlier cutoff timestamp) and we'll hopefully have fixed nonces to use transients or something better than `wp_options` by 2286, we can probably live with this potential bug for now.

sergeymitr replied:
> we'll hopefully have fixed nonces to use transients or something better than wp_options by 2286

The goal worthy of a noble man 😀

anomiex commented on packages/connection/tests/php/test-class-nonce-handler.php:
```diff
@@ -0,0 +1,178 @@
+	public function test_add_existing() {
+		$query_filter_run = false;
+
+		$nonce_object = array(
```
Should you cast to object here? None of the code currently cares what's actually returned from the query, but it'd be nice to be accurate.

sergeymitr replied:
Done 👍

anomiex reviewed (commented):
I see no major issues. I made several suggestions, but nothing I feel strongly enough about to try to block an approval. Several are questions because I'm still learning.

sergeymitr committed:
API Nonces: Minor refactoring.

sergeymitr commented on class.jetpack.php:
```diff
@@ -3288,9 +3294,13 @@ public static function plugin_deactivation() {
 	public static function disconnect( $update_activated_state = true ) {
+		// The hook is not being set since Jetpack 9.0.0,
+		// but we're removing it just in case it wasn't properly cleaned up after the plugin update.
 		wp_clear_scheduled_hook( 'jetpack_clean_nonces' );
```
Good point, that's unnecessary.
Removed the condition.

sergeymitr commented on class.jetpack.php:
```diff
@@ -5553,11 +5563,7 @@ public function wp_rest_authentication_errors( $value ) {
 	public function add_nonce( $timestamp, $nonce ) {
 		_deprecated_function( __METHOD__, 'jetpack-7.7', 'Automattic\\Jetpack\\Connection\\Manager::add_nonce' );
```
Yes, I believe we can remove it, good catch.

sergeymitr commented on packages/connection/src/class-nonce-handler.php:
```diff
@@ -0,0 +1,197 @@
+			return static::$nonces_used_this_request[ "$timestamp:$nonce" ];
+		}
+
+		// This should always have gone through Jetpack_Signature::sign_request() first to check $timestamp an $nonce.
```
Wow, that type has been there since the beginning of Jetpack: https://github.com/Automattic/jetpack/commit/99e10babcb00e4d5f9a8d9d91584f71626fceb2f
Literally the first commit 😀

sergeymitr commented on packages/connection/src/class-nonce-handler.php:
```diff
@@ -0,0 +1,197 @@
+		$nonce     = esc_sql( $nonce );
+
+		// Raw query so we can avoid races: add_option will also update.
+		$show_errors = $wpdb->show_errors( false );
```
Fixed 👍

sergeymitr commented on packages/connection/src/class-nonce-handler.php:
```diff
@@ -0,0 +1,197 @@
+		$show_errors = $wpdb->show_errors( false );
+
+		$old_nonce = $wpdb->get_row(
+			$wpdb->prepare( "SELECT * FROM `$wpdb->options` WHERE option_name = %s", "jetpack_nonce_{$timestamp}_{$nonce}" )
```
Fixed 👍

anomiex commented on packages/connection/tests/php/test-class-nonce-handler.php:
```diff
@@ -86,7 +86,7 @@ public function test_add_existing() {
-				$this->assertEquals( "SELECT * FROM `options` WHERE option_name = '{$nonce_object['option_name']}'", $query );
+				$this->assertEquals( "SELECT 1 FROM `options` WHERE option_name = '{$nonce_object->option_name}'", $query );
 
 				return array( $nonce_object );
```
Now with the changed query, the return should be something along the lines of

`return array( (object) array( 1 => '1' ) );`

to match what the actual DB will return.

sergeymitr replied:
Fixed 👍

sergeymitr committed:
API Nonces: More refactoring.

sergeymitr commented on packages/connection/src/class-nonce-handler.php:
```diff
@@ -0,0 +1,205 @@
+				$return = false;
+			}
+		} finally {
+			$show_errors && $wpdb->show_errors();
```
Good catch, fixed.

sergeymitr committed:
API Nonces: Time limit on full cleanup.

Having too many nonces to remove can break Jetpack disconnecting process. We limit the cleanup to 20 seconds and 1 million records, whichever reached earlier.

anomiex commented on packages/connection/src/class-nonce-handler.php:
```diff
@@ -0,0 +1,206 @@
+
+	/**
+	 * Removing [almost] all the nonces.
+	 * Capped at 1 million records to avoid breaking the site.
```
Update the comment to refer to the 20-second time limit?

anomiex commented on packages/connection/src/class-nonce-handler.php:
```diff
@@ -0,0 +1,206 @@
+	 */
+	public static function clean_all( $cutoff_timestamp = PHP_INT_MAX ) {
+		// phpcs:ignore Generic.CodeAnalysis.ForLoopWithTestFunctionCall.NotAllowed
+		for ( $start_time = time(), $i = 0; $i < 1000 && time() < $start_time + 20; ++$i ) {
```
Since we have a time limit now, should we drop the arbitrary "1000 batches" limit?

Also, it's a microoptimization, but we could avoid the addition on each iteration of the loop with something like
```
for ( $end_time = time() + 20; time() < $end_time; ) {
```

anomiex reviewed (commented):
Could use a doc comment update now. Code seems fine, although I left a suggestion for improvement.

sergeymitr committed:
API Nonces: Removing 1000-batch limit from full cleanup.

The full nonce cleanup is only capped at 20 seconds of time.
There's no longer any need to cap it at 1 million nonces.

anomiex reviewed (dismissed):
Code seems fine. Travis seems to be unhappy though.

kraftbj commented:
A note on Travis. The passing runs are symlinking, the failing runs are grabbing from git.

anomiex commented:
> Looking in setup-travis.sh, we need to remove phpcs-changed from the dev list to pass PHP minimum version. That's may be breaking the composer.lock and having it use a git version, instead of symlinking?

I think I figured out what's going on here, and what changed recently to suddenly break it.

`composer.lock` currently specifies the versions of all the packages from the monorepo as "dev-master". This was just changed yesterday, by #17101 which changed them all from "dev-add/jetpack-lazy-images-package".

A normal `composer install` apparently just blindly trusts the sources listed `composer.lock`, ignoring the versions specified there at least for the monorepo. So it uses the monorepo, and everything is fine.

A normal `composer update` takes the first matching version from the first repository that has one. Since we declare "@dev", the version in the monorepo matches and that's what it uses.

It turns out that the `composer remove sirbrillig/phpcs-changed --dev` is implicitly doing a `composer update sirbrillig/phpcs-changed` (maybe with `--with-dependencies`). The specification of the package name tells the update that only that package should be updated, everything else should not be changed. When packages are already installed it seems to trust that whatever is on disk is accurate, but if they're not already installed it decides it has to do so. And it seems to do this by looking for a version matching that in `composer.lock` in all known repos first _before_ trusting the source specified in `composer.lock`. Since packagist does have "dev-master" while it considers the monorepo as being something like "dev-ef396149805c6d17f2bac0aff8db59229ab18a6e", it goes with packagist. It didn't break before #17101 because packagist doesn't have any version for "dev-add/jetpack-lazy-images-package", so it fell back to trusting `composer.lock` when it said "dev-master" was the monorepo path.

anomiex commented:
As for fixing it, I found a few options:

* Just change composer.lock back to specifying something other than "dev-master", so it doesn't see packagist.org as having a matching version. And maybe add a test or hook or something to prevent `composer.lock` from referring to the monorepo with "dev-master" as the version.
* Before the `composer remove`, do `composer install --ignore-platform-reqs` to get files on-disk.
* Add `--no-update` to `composer remove` so it only updates `composer.json`, then do `composer update` (with no list of packages) so it pulls from the monorepo.

kraftbj committed:
Fix composer.lock issue

kraftbj commented:
Did:

```
git checkout f0c54b8942f82f182890a1ac86c6f06bd796a85f composer.lock
composer update automattic/jetpack-blocks
```

to reset the status of the composer.lock file.

cc: @jeherve
